Memory viz cli

sarahsonder opened the pull request:
## Proposed Changes

The purpose of this pull request is to create a command line interface for MemoryViz. This is to make it easier to run MemoryViz.

To run the CLI, run `npm i` to install the CLI and then run `memory-viz <file-to-path>`, replacing `<path-to-file>` with the path to a file containing MemoryViz-compatible JSON

## Type of Change

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |    X     |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist


Before opening your pull request:

-   [X] I have performed a self-review of my changes.
    -   Check that all changed files included in this pull request are intentional changes.
    -   Check that all changes are relevant to the purpose of this pull request, as described above.
-   [X] I have added tests for my changes, if applicable.
    -   This is **required** for all bug fixes and new features.
-   [X] I have updated the project documentation, if applicable.
    -   This is **required** for new features.
-   [X] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

-   [x] I have verified that the CI tests have passed.
-   [x] I have reviewed the test coverage changes reported by Coveralls.
-   [x] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

1. I am unsure as the why the Github tests are failing. The issue seems to be that the files/branch cannot be found. I left a more detailed message on Slack. Thank you!
2. Since my current implementation takes in only one argument, which is the path to the file containing the Memory-Viz compatible JSON input, users do not have a choice as to where the SVG image gets saved. This also means that users cannot choose the name of the output, which means that old SVGs could get overwritten. Would you like me to change the implementation such that users may specify where they would like the image to be saved? For reference, I currently have it so that the image gets saved in the same directory in which Memory-Viz is being called. I can also look into renaming files with (1) or (2) if files of the same name already exist in a directory.
3. Would you like me to create a new page in the docs website to document the CLI or should I do it elsewhere?
4. Is there a particular library that you would like me to use for testing the CLI?
5. Are we only allowing JSON files as inputs for the moment?

david-yz-liu commented:
@sarahsonder good questions.

1. For the output file: the current directory is fine. For the filename, take the base name of the input json and change the extension to `.svg`. (Future work could involve allowing the user to specify an output file.) 
2. For documentation: (1) add a brief subsection to the `memory-viz/README.md` with a sample invocation, and (2) yes create a new documentation page for the CLI. It can be pretty bare-bones for now, but hopefully it will grow over time as we add new options. 😄 You can use some professional pages as templates to follow, e.g. https://jestjs.io/docs/cli. 
3. You should stick with Jest testing. For now you can manually create a subprocess to execute an `npx memory-viz <whatever>` and check the file contents (ideally, against a snapshot!).
4. Yes that's correct, though I'm not sure what other file types you're thinking of.

sarahsonder commented:
## Proposed Changes

The purpose of this pull request is to create a command line interface for MemoryViz. This is to make it easier to run MemoryViz.

To run the CLI, run `npm i` to install the CLI and then run `npx memory-viz <file-to-path>`, replacing `<path-to-file>` with the path to a file containing MemoryViz-compatible JSON

## Type of Change

| Type                                                                                    | Applies? |
| --------------------------------------------------------------------------------------- | -------- |
| 🚨 _Breaking change_ (fix or feature that would cause existing functionality to change) |          |
| ✨ _New feature_ (non-breaking change that adds functionality)                          |    X     |
| 🐛 _Bug fix_ (non-breaking change that fixes an issue)                                  |          |
| 🎨 _User interface change_ (change to user interface; provide screenshots)              |          |
| ♻️ _Refactoring_ (internal change to codebase, without changing functionality)          |          |
| 🚦 _Test update_ (change that _only_ adds or modifies tests)                            |          |
| 📚 _Documentation update_ (change that _only_ updates documentation)                    |          |
| 📦 _Dependency update_ (change that updates a dependency)                               |          |
| 🔧 _Internal_ (change that _only_ affects developers or continuous integration)         |          |

## Checklist


Before opening your pull request:

-   [X] I have performed a self-review of my changes.
    -   Check that all changed files included in this pull request are intentional changes.
    -   Check that all changes are relevant to the purpose of this pull request, as described above.
-   [X] I have added tests for my changes, if applicable.
    -   This is **required** for all bug fixes and new features.
-   [X] I have updated the project documentation, if applicable.
    -   This is **required** for new features.
-   [X] I have updated the project Changelog (this is required for all changes).

After opening your pull request:

-   [ ] I have verified that the CI tests have passed.
-   [ ] I have reviewed the test coverage changes reported by Coveralls.
-   [ ] I have [requested a review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) from a project maintainer.

## Questions and Comments

I am unsure as the why the Github tests are failing. The issue seems to be that the files/branch cannot be found. I left a more detailed message on Slack. Thank you!

david-yz-liu commented:
@sarahsonder don't put the pull request template in a comment. Instead edit the pull request description so that it contains the full information.

david-yz-liu commented on CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ and adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 -   Removed the `isClass` and `stack_frame` attributes and embedded them as the types `.class` and `.frame`.
 -   Renamed the input for blank objects from `BLANK` to `.blank`.
 -   Created new type `.blank-frame` to denote blank stack frames.
+-   Created a CLI for MemoryViz.
```
This is an "Enhancement" because it just adds new functionality.

david-yz-liu commented on README.md:
````diff
@@ -52,6 +52,16 @@ produces a file `simple_demo.svg` that looks like the following:
+$ npx memory-viz <path-to-file>
+```
+
+replacing `<path-to-file>` with the path to a file containing MemoryViz-compatible JSON. The output is an SVG image generated by MemoryViz and the image is saved in the same folder in which the CLI is called. For more information, check out the project [documentation website](https://www.cs.toronto.edu/~david/memory-viz/docs/cli).
````
"in the same folder in which the CLI is called" -> "in the current working directory"

(same wording change across the other documentation)

david-yz-liu commented on memory-viz/bin/cli.js:
```diff
@@ -0,0 +1,54 @@
+// Check for correct number of arguments
+if (process.argv.length !== 3) {
+    console.error(
+        "Error: wrong number of arguments.\nProper input: npx memory-viz <path-to-file>"
```
"Proper input" -> "Usage"

More importantly, `npx` is only necessary when the package is not installed. When the package is already installed, `memory-viz <path-to-file>` should suffice. Please test that out, as it does have implications for the other sections (and testing)

david-yz-liu commented on memory-viz/bin/cli.js:
```diff
@@ -0,0 +1,54 @@
+if (!fs.existsSync(absolutePath)) {
+    console.error(`Error: File ${absolutePath} does not exist.`);
+    process.exit(1);
+} else if (path.extname(filePath) !== ".json") {
```
Don't make this check (e.g. a `.txt` file might contain valid JSON)

david-yz-liu commented on memory-viz/bin/cli.js:
```diff
@@ -0,0 +1,54 @@
+
+let m;
+try {
+    m = draw(data, true, { width: 1300, seed: 12345 });
```
Add a TODO comment here to replace `width` and `seed` with command-line arguments in the future

david-yz-liu commented on memory-viz/bin/cli.js:
```diff
@@ -0,0 +1,54 @@
+    process.exit(1);
+}
+
+const outputName = path.basename(filePath).slice(0, -5) + ".svg";
```
Use [`path.parse`](https://nodejs.org/api/path.html#pathparsepath) to obtain the filename without the extension

david-yz-liu commented on memory-viz/src/tests/draw.spec.tsx:
```diff
@@ -1,5 +1,121 @@
 import exports from "../index";
 const { MemoryModel, draw } = exports;
+const { exec, execSync } = require("child_process");
```
Add these tests to a separate file (perhaps `cli.spec.tsx`)

david-yz-liu reviewed (commented):
@sarahsonder good work. Regarding the failing tests, there's an error in how the module is being imported (and likely your local environment is different from GitHub Actions).

Try using a different approach:

1. Modify the GitHub Actions workflow to first run the webpack build command before running the tests
2. (perhaps try this, not sure) In the test cases, invoke just `memory-viz` rather than `npx memory-viz`

david-yz-liu commented on .github/workflows/ci.yml:
```diff
@@ -40,7 +40,9 @@ jobs:
             - name: Run jest tests
-              run: npm run test-cov
+              run: |
+                  npm run build --workspace=memory-viz
```
Make this a separate step preceding "Run jest tests"

david-yz-liu commented on docs/docs/06-cli.md:
````diff
@@ -0,0 +1,21 @@
+$ npx memory-viz <path-to-file>
+```
+
+replacing `<path-to-file>` with the path to a file containing MemoryViz-compatible JSON. If the file extension is not `.json` or if the file content is not compatible with MemoryViz, an error will be thrown.
````
We're no longer checking file extension

david-yz-liu commented on memory-viz/bin/cli.js:
```diff
@@ -0,0 +1,55 @@
+try {
+    data = JSON.parse(fileContent);
+} catch (err) {
+    console.error(`Error: Invalid JSON.`);
```
Print out the error message here as well
